# Oct, 2021

## 2021-10-22: Megalodon rf


## 2021-10-22: Megalodon rf


## 2021-10-22: Megalodon rf


## 2021-10-21: Megalodon rf

Build 2 pipelines:

1. RandomUnderSampler -> SMOTE -> random forest(Random search): `p1_rf_RandomUnderSampler_SMOTE_RandomCV.py`

2. ENN -> SMOTE -> random forest(Random search): `p2_rf_ENN_SMOTE_RandomCV.py`

## 2021-10-19: Megalodon rf

1. Modify the confusin matrix script `2021-10-19/plot_confusion_matrix.py`

2. Explore the strategy to combine undersampling with Edited Nearest Neighbors and oversampling with SMOTE

## 2021-10-18: model build

Apply extract feature module for T4_5hmC/lambda_5mC/lambda_5C fast5 files(Guppy v5.0.11+Tombo 1.5.1)

Result: `/pod/2/li-lab/Ziwei/Nanopore/results/feature_Guppy5.0.11/`

## 2021-10-17: model build

Extract feature module debugging:

Modify the code from 2021-10-12: delete multiprocess.queue() at this moment since it didn't work. Right now it can only used single process without batch size

## 2021-10-14: Other

Test the DeepSignal on its speed and output for its feature extraction module

## 2021-10-13: Megalodon rf

Write a functional optimized script(`rf_pipeline_gridsearch.py`) for SMOTE + GridSearch for random forest model for Megalodon
```shell
script_dir=/pod/2/li-lab/Ziwei/Nanopore/daily/2021-10-13
#input_file=/pod/2/li-lab/Ziwei/Nanopore/daily/test/total.test.bed.gz
input_file='/pod/2/li-lab/Ziwei/Nanopore/results/megalodon2.3.4_guppy5.0.14/total.Megalodon.per_read.prob.bed.gz'
python $script_dir/rf_pipeline_gridsearch.py --input_path $input_file --output_path $script_dir > $script_dir/rf_params_smote_gridsearch.txt
```

## 2021-10-12: model build

Script to extract features from fast5 files: `cytosine/extract_module.py`

Signal feature: 1. kmer sequence 2. signal mean for each base in kmer 3. signal std for each base in kmer 4.signal number for each base in kmer 5. signal range for each base in kmer

## 2021-10-11: Megalodon rf

SMOTE + GridSearch for random forest model for Megalodon

## 2021-10-08: Megalodon rf
1. SMOTE + RandomSearch for random forest model for Megalodon

The model with best parameter is saved in `rf.pkl`

I changed the strategy to escape data leakage and follow the steps below(apply SMOTE only on the training dataset, not the whole dataset):
1. Split the sample into training and testing datasets
2. SMOTE to oversample the training dataset
3. Train the the classifier on the training fold for random forest 
4. Validate the classifer on the reamining fold
5. Save the classification report and confusion matrix plotting for default parameter

2. Write a script `read_pkl.py` to read the best model(.pkl)

3. Compare the fast5 files after Guppy-basecalling amd Tombo-requigglling for Guppy v5.0.11/v5.0.14 + Tombo 1.5.1

Conclusion: The fast5 number is smaller in Guppy v5.0.14 than Guppy v5.0.11. T4 is detected with Guppy v5.0.14, I used Guppy v5.0.11 instead.

See result/fast5_count.csv for the exact number

## 2021-10-07: Process

Redo the Guppy-basecalling amd Tombo-requigglling with 1)Guppy v5.0.11 + Tombo 1.5.1  2)Guppy v5.0.14 + Tombo 1.5.1


## 2021-10-05: Process
1. Explore the signal feature after Guppy-basecalling Tombo-requiggled fast5 files: `feature_100521.ipynb`

2. Write `fast5_parser.py` to extract raw signal, event, alignment information from single basecalled-resquirred fast5 read (Exploration: `feature_script.ipynb`)

3. Write `PerChromSeperation.py` to get the list of fast5 files based on chromosome information

